### DISCUSSION TITLE
showing a different set of stimuli for every participant

### DISCUSSION THREAD

**Discussion by veranicolette** [jspsych/jsPsych#1187] [Q&A] [answered]
I'm having trouble trying to figure out how to present different sets of stimuli for my experiment. Right now, I have a JSON file containing 300 sentences and I am pulling 150 sentences from it using sample without replacement. This is what that portion of my code currently looks like:

`var trial = {
            type: "survey-html-form",
            html: jsPsych.timelineVariable('SENTENCE'),
        }

        var print_out_sentences = {
            timeline: [trial],
            timeline_variables: sample_sentences,
            sample: {
                type: "without-replacement",
                size: 150,
            },
            randomize_order: true,
            data: {given_sentence: jsPsych.timelineVariable('SENTENCE')},
            on_finish: function(data) {
                    count++;
                    var progress = count/n_trials;
                    jsPsych.setProgressBar(progress);
            }

        }
        timeline.push(print_out_sentences);`

My PI wants me to create 2 versions of stimuli with 150 sentences each (which I already have) so the sentences have a somewhat equal probability of being shown to the participant, but I am not sure how to implement that.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @veranicolette, if you want to randomly select between the two versions each time the experiment runs, you could do it like this:
```javascript
// create a variable to hold the stimuli information to be used for timeline variables
// this will be one of the two sets of 150 sentences
var sentences;
// randomly select a version: 1 or 2
var version = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
// use the version number to assign one of the two sentence arrays to the 'sentences' variable
if (version == 1) {
  sentences = sentence_set_1;
} else {
  sentences = sentence_set_2;
}
// set up your trial procedure using the 'sentences' variable as your timeline_variables array
var print_out_sentences = {
  timeline: [trial],
  timeline_variables: sentences,
  // etc.
}
```

**Reply by veranicolette** [reply]
Hi @becky-gilbert, I tried it and my experiment won't run. The console says what I inputted in sentences, stimuli_list1 and stimuli_list2 aren't defined. Both files have been loaded at the top though along with the plugins etc.

**Reply by veranicolette** [reply]
This is what my code looks like right now
`var sentences;
        var version = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
        if (version == 1){
            sentences = stimuli_list1;
        } else {
            sentences = stimuli_list2;
        }

        var trial = {
            type: "survey-html-form",
            html: jsPsych.timelineVariable('SENTENCE'),
        }

        var print_out_sentences = {
            timeline: [trial],
            timeline_variables: sentences,
            randomize_order: true,
            data: {given_sentence: jsPsych.timelineVariable('SENTENCE')},
            on_finish: function(data) {
                    count++;
                    var progress = count/n_trials;
                    jsPsych.setProgressBar(progress);
            }

        }
        timeline.push(print_out_sentences);`

**Reply by jodeleeuw** [reply]
Can you post your HTML code and the stimuli files as well, please?

**Reply by veranicolette** [reply]
Here's my code:
`<!DOCTYPE html>
<html>
    <head>
        <title> Lab Experiment </title>
        <script src = "jspsych-6.1.0/jspsych.js"></script>
        <script src = "jspsych-6.1.0/plugins/jspsych-survey-html-form.js"></script>
        <script src = "jspsych-6.1.0/plugins/jspsych-html-button-response.js"></script>
        <script src = "jspsych-6.1.0/plugins/jspsych-html-keyboard-response.js"></script>
        <script src = "shorter_stimuli.js"></script>
        <script src = "stimuli.js"></script>
        <script src = "stimuli_list1.js"></script>
        <script src = "stimuli_list2.js"></script>
        <script src = "consent/consent_form.pdf"></script>
        <link href = "jspsych-6.1.0/css/jspsych.css" rel = "stylesheet">
    </head> 

    <body></body>    
    <script>

        var timeline = [];

        jsPsych.data.addProperties({
            subject_id: jsPsych.randomization.randomID(15),
        });

        var consent = {
            type: "html-button-response",
            stimulus: '<iframe src = "consent/consent_form.pdf" style="width:1000px; height:800px;" frameborder="0" allowfullscreen></iframe> ' + '<br></br>',
            choices: ["I HAVE READ THE PRECEDING INFORMATION AND WOULD LIKE TO PARTICIPATE IN THE STUDY", "I WOULD LIKE TO WITHDRAW FROM THIS STUDY. (There is no penalty for withdrawing.)"],
            
            on_finish: function(data) {
                if (data.button_pressed == 1) {
                jsPsych.endExperiment("Thank you for your response. <p> Please return this HIT through the Amazon Mechanical Turk platform (you may close this browser window). </p>");
                }
            }
        };
        timeline.push(consent);

        
        var count = 0;
        var n_trials = 150;

        var sentences;
        var version = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
        if (version == 1){
            sentences = stimuli_list1;
        } else {
            sentences = stimuli_list2;
        }

        var trial = {
            type: "survey-html-form",
            html: jsPsych.timelineVariable('SENTENCE'),
        }

        var print_out_sentences = {
            timeline: [trial],
            timeline_variables: sentences,
            randomize_order: true,
            data: {given_sentence: jsPsych.timelineVariable('SENTENCE')},
            on_finish: function(data) {
                    count++;
                    var progress = count/n_trials;
                    jsPsych.setProgressBar(progress);
            }

        }
        timeline.push(print_out_sentences);

        var code = {
            type: "html-keyboard-response",
            choices: jsPsych.NO_KEYS,
            stimulus: function(){
            return "You have completed the main task! Please copy the following code to the MTurk page. You can then close this window.  <br>" + jsPsych.data.get().values()[0].subject_id + "<br> Your browser may ask you to confirm that you want to to leave this page, just say leave, we have recorded your performance."
            }
        }
        timeline.push(code);

    
        jsPsych.init({
            timeline: timeline,
            show_progress_bar: true,
            auto_update_progress_bar: false,
            
        })


    </script>
    
</html>

`

**Reply by veranicolette** [reply]
I can't upload my stimuli files and they're pretty long so I'll just copy part of one file if that's okay. They both look the same, they just differ in the sentences that are in the file. I worked off of your video on how to load data from csv files when I was first working on this and that worked just fine when I was only pulling the sentences from one csv file that I converted into a JSON.

This is what's in stimuli_list1.js:

var sample_sentences = [
    {
      "SENTENCE": "<p> Ruby's grandchildren hired a live-in nurse because they thought she was too old to live <input type = 'text' name = 'user_response'></p>"
    },
    
    {
      "SENTENCE": "<p> He wrote himself a note so that he wouldn't <input type = 'text' name = 'user_response'></p>"
    },
    
    {
      "SENTENCE": "<p> Every winter you see the Hernandez house and its huge chimney with puffs of billowing <input type = 'text' name = 'user_response'></p>"
    },
    
    {
      "SENTENCE": "<p> Marsha wasn't worried about dinner because she knew the guests could remain eating appetizers for at least another <input type = 'text' name = 'user_response'></p>"
    },
    
    {
      "SENTENCE": "<p> The little puppy grew up to be a huge <input type = 'text' name = 'user_response'></p>"
    },

    {
      "SENTENCE": "<p> I  added my name to the <input type = 'text' name = 'user_response'></p>"
    },

    {
      "SENTENCE": "<p> The teacher saw that some students were cheating and wrote down their <input type = 'text' name = 'user_response'></p>"
    },
]

**Reply by veranicolette** [reply]
The first sentence has the same tags as the others, I'm not sure why it keeps disappearing when I copy it on here

**Reply by becky-gilbert** [reply]
Hi @veranicolette, can you try renaming your two sentence arrays to `stimuli_list1` and `stimuli_list2`. So for instance, inside your stimuli_list1.js file, it would look like this:
```javascript
var stimuli_list1 = [
  {"SENTENCE": "Ruby's grandchildren hired a live-in nurse because they thought she was too old to live"},
  // etc.
];
```
And the same for stimuli_list2.js. That should give you access to the `stimuli_list1` and `stimuli_list2` arrays inside your main script, so that you can assign one of them to your `sentences` array, depending on the version that was selected:
```javascript
if (version == 1){
  sentences = stimuli_list1;
} else {
  sentences = stimuli_list2;
}
```

**Reply by veranicolette** [reply]
I renamed the sentence arrays and the experiment is working now! Thank you so much for your help!
